File: bcmc-ui/src/main/java/com/adyen/checkout/bcmc/BcmcView.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by arman on 18/9/2019.
 */

package com.adyen.checkout.bcmc;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import android.content.Context;
import android.content.res.TypedArray;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.textfield.TextInputLayout;
import android.text.Editable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;

import com.adyen.checkout.base.PaymentComponentState;
import com.adyen.checkout.base.api.ImageLoader;
import com.adyen.checkout.base.ui.view.AdyenLinearLayout;
import com.adyen.checkout.base.ui.view.AdyenTextInputEditText;
import com.adyen.checkout.base.ui.view.RoundCornerImageView;
import com.adyen.checkout.base.validation.ValidatedField;
import com.adyen.checkout.bcmc.ui.R;
import com.adyen.checkout.card.CardValidationUtils;
import com.adyen.checkout.card.ui.CardNumberInput;
import com.adyen.checkout.card.ui.ExpiryDateInput;

/**
 * CardView for {@link BcmcComponent}.
 */
@SuppressWarnings("SyntheticAccessor")
public final class BcmcView
        extends AdyenLinearLayout<BcmcOutputData, BcmcConfiguration, PaymentComponentState, BcmcComponent> implements Observer<BcmcOutputData> {

    private RoundCornerImageView mCardBrandLogoImageView;

    private CardNumberInput mCardNumberEditText;
    private ExpiryDateInput mExpiryDateEditText;

    private TextInputLayout mExpiryDateInput;
    private TextInputLayout mCardNumberInput;

    private final BcmcInputData mCardInputData = new BcmcInputData();

    private ImageLoader mImageLoader;

    public BcmcView(@NonNull Context context) {
        this(context, null);
    }

    public BcmcView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    // Regular View constructor
    @SuppressWarnings("JavadocMethod")
    public BcmcView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setOrientation(LinearLayout.VERTICAL);

        LayoutInflater.from(getContext()).inflate(R.layout.bcmc_view, this, true);

        final int padding = (int) getResources().getDimension(R.dimen.standard_margin);
        setPadding(padding, padding, padding, 0);
    }

    @Override
    protected void initLocalizedStrings(@NonNull Context localizedContext) {
        final int[] myAttrs = {android.R.attr.hint};
        TypedArray typedArray;

        // Card Number
        typedArray = localizedContext.obtainStyledAttributes(R.style.AdyenCheckout_Card_CardNumberInput, myAttrs);
        mCardNumberInput.setHint(typedArray.getString(0));
        typedArray.recycle();

        // Expiry Date
        typedArray = localizedContext.obtainStyledAttributes(R.style.AdyenCheckout_Card_ExpiryDateInput, myAttrs);
        mExpiryDateInput.setHint(typedArray.getString(0));
        typedArray.recycle();
    }

    @Override
    public void initView() {
        mCardBrandLogoImageView = findViewById(R.id.cardBrandLogo_imageView);

        initCardNumberInput();
        initExpiryDateInput();
    }

    @Override
    public void onComponentAttached() {
        mImageLoader = ImageLoader.getInstance(getContext(), getComponent().getConfiguration().getEnvironment());
    }

    @Override
    public void onChanged(@Nullable BcmcOutputData cardOutputData) {
        if (cardOutputData != null) {
            onCardNumberValidated(cardOutputData.getCardNumberField());
        }
    }

    @Override
    protected void observeComponentChanges(@NonNull LifecycleOwner lifecycleOwner) {
        getComponent().observeOutputData(lifecycleOwner, this);
    }

    @Override
    public boolean isConfirmationRequired() {
        return true;
    }

    @Override
    public void highlightValidationErrors() {
        final BcmcOutputData outputData;
        if (getComponent().getOutputData() != null) {
            outputData = getComponent().getOutputData();
        } else {
            return;
        }

        boolean isErrorFocused = false;

        if (!outputData.getCardNumberField().isValid()) {
            isErrorFocused = true;
            mCardNumberEditText.requestFocus();
            mCardNumberInput.setError(mLocalizedContext.getString(R.string.checkout_card_number_not_valid));
        }

        if (!outputData.getExpiryDateField().isValid()) {
            if (!isErrorFocused) {
                mExpiryDateInput.requestFocus();
            }
            mExpiryDateInput.setError(mLocalizedContext.getString(R.string.checkout_expiry_date_not_valid));
        }
    }

    private void notifyInputDataChanged() {
        getComponent().inputDataChanged(mCardInputData);
    }

    private void onCardNumberValidated(@NonNull ValidatedField<String> validatedNumber) {
        if (validatedNumber.getValidation() == ValidatedField.Validation.VALID) {
            changeFocusOfInput(validatedNumber.getValue());
        }

        if (!getComponent().isCardNumberSupported(validatedNumber.getValue())) {
            mCardBrandLogoImageView.setStrokeWidth(0f);
            mCardBrandLogoImageView.setImageResource(R.drawable.ic_card);
        } else {
            mCardBrandLogoImageView.setStrokeWidth(RoundCornerImageView.DEFAULT_STROKE_WIDTH);
            mImageLoader.load(BcmcComponent.SUPPORTED_CARD_TYPE.getTxVariant(), mCardBrandLogoImageView);
        }
    }

    private void changeFocusOfInput(String numberValue) {
        final int length = numberValue.length();

        if (length == CardValidationUtils.MAXIMUM_CARD_NUMBER_LENGTH) {
            goToNextInputIfFocus(mCardNumberEditText);
        }
    }

    private void goToNextInputIfFocus(View view) {
        if (getRootView().findFocus() == view) {
            findViewById(view.getNextFocusForwardId()).requestFocus();
        }
    }

    private void initCardNumberInput() {
        mCardNumberInput = findViewById(R.id.textInputLayout_cardNumber);
        mCardNumberEditText = (CardNumberInput) mCardNumberInput.getEditText();
        //noinspection ConstantConditions
        mCardNumberEditText.setOnChangeListener(new AdyenTextInputEditText.Listener() {
            @Override
            public void onTextChanged(@NonNull Editable editable) {
                mCardInputData.setCardNumber(mCardNumberEditText.getRawValue());
                notifyInputDataChanged();
                mCardNumberInput.setError(null);
            }
        });
        mCardNumberEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                final BcmcOutputData outputData = getComponent().getOutputData();
                if (hasFocus) {
                    mCardNumberInput.setError(null);
                } else if (outputData != null && !outputData.getCardNumberField().isValid()) {
                    mCardNumberInput.setError(mLocalizedContext.getString(R.string.checkout_card_number_not_valid));
                }
            }
        });
    }

    private void initExpiryDateInput() {
        mExpiryDateInput = findViewById(R.id.textInputLayout_expiryDate);
        mExpiryDateEditText = (ExpiryDateInput) mExpiryDateInput.getEditText();
        //noinspection ConstantConditions
        mExpiryDateEditText.setOnChangeListener(new AdyenTextInputEditText.Listener() {
            @Override
            public void onTextChanged(@NonNull Editable editable) {
                mCardInputData.setExpiryDate(mExpiryDateEditText.getDate());
                notifyInputDataChanged();
                mExpiryDateInput.setError(null);
            }
        });
        mExpiryDateEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                final BcmcOutputData outputData = getComponent().getOutputData();
                if (hasFocus) {
                    mExpiryDateInput.setError(null);
                } else if (outputData != null && !outputData.getExpiryDateField().isValid()) {
                    mExpiryDateInput.setError(mLocalizedContext.getString(R.string.checkout_expiry_date_not_valid));
                }
            }
        });
    }
}

File: bcmc/src/main/java/com/adyen/checkout/bcmc/internal/ui/view/BcmcView.kt
/*
 * Copyright (c) 2022 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by oscars on 29/9/2022.
 */

package com.adyen.checkout.bcmc.internal.ui.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.widget.CompoundButton
import android.widget.LinearLayout
import androidx.annotation.StringRes
import androidx.core.view.isVisible
import com.adyen.checkout.bcmc.R
import com.adyen.checkout.bcmc.databinding.BcmcViewBinding
import com.adyen.checkout.bcmc.internal.ui.BcmcDelegate
import com.adyen.checkout.components.base.ComponentDelegate
import com.adyen.checkout.components.extensions.hideError
import com.adyen.checkout.components.extensions.isVisible
import com.adyen.checkout.components.extensions.setLocalizedHintFromStyle
import com.adyen.checkout.components.extensions.setLocalizedTextFromStyle
import com.adyen.checkout.components.extensions.showError
import com.adyen.checkout.components.ui.ComponentView
import com.adyen.checkout.components.ui.Validation
import kotlinx.coroutines.CoroutineScope

internal class BcmcView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr),
    ComponentView {

    private val binding = BcmcViewBinding.inflate(LayoutInflater.from(context), this)

    private lateinit var localizedContext: Context

    private lateinit var delegate: BcmcDelegate

    init {
        orientation = VERTICAL
        val padding = resources.getDimension(R.dimen.standard_margin).toInt()
        setPadding(padding, padding, padding, 0)
    }

    override fun initView(delegate: ComponentDelegate, coroutineScope: CoroutineScope, localizedContext: Context) {
        if (delegate !is BcmcDelegate) throw IllegalArgumentException("Unsupported delegate type")
        this.delegate = delegate

        this.localizedContext = localizedContext
        initLocalizedStrings(localizedContext)

        initCardNumberInput()
        initExpiryDateInput()
        initCardHolderInput()
        initStorePaymentMethodSwitch()
    }

    private fun initLocalizedStrings(localizedContext: Context) {
        with(binding) {
            textInputLayoutCardNumber.setLocalizedHintFromStyle(
                R.style.AdyenCheckout_Card_CardNumberInput,
                localizedContext
            )
            textInputLayoutExpiryDate.setLocalizedHintFromStyle(
                R.style.AdyenCheckout_Card_ExpiryDateInput,
                localizedContext
            )
            binding.textInputLayoutCardHolder.setLocalizedHintFromStyle(
                R.style.AdyenCheckout_Card_HolderNameInput,
                localizedContext
            )
            switchStorePaymentMethod.setLocalizedTextFromStyle(
                R.style.AdyenCheckout_Card_StorePaymentSwitch,
                localizedContext
            )
        }
    }

    private fun initExpiryDateInput() {
        binding.editTextExpiryDate.setOnChangeListener {
            delegate.updateInputData { expiryDate = binding.editTextExpiryDate.date }
            binding.textInputLayoutExpiryDate.hideError()
        }

        binding.editTextExpiryDate.onFocusChangeListener = OnFocusChangeListener { _: View?, hasFocus: Boolean ->
            val expiryDateValidation = delegate.outputData.expiryDateField.validation
            if (hasFocus) {
                binding.textInputLayoutExpiryDate.hideError()
            } else if (expiryDateValidation is Validation.Invalid) {
                val errorReasonResId = expiryDateValidation.reason
                binding.textInputLayoutExpiryDate.showError(localizedContext.getString(errorReasonResId))
            }
        }
    }

    private fun initCardNumberInput() {
        binding.editTextCardNumber.setOnChangeListener {
            delegate.updateInputData { cardNumber = binding.editTextCardNumber.rawValue }
            setCardNumberError(null)
        }

        binding.editTextCardNumber.onFocusChangeListener = OnFocusChangeListener { _: View?, hasFocus: Boolean ->
            val cardNumberValidation = delegate.outputData.cardNumberField.validation
            if (hasFocus) {
                setCardNumberError(null)
            } else if (cardNumberValidation is Validation.Invalid) {
                val errorReasonResId = cardNumberValidation.reason
                setCardNumberError(errorReasonResId)
            }
        }
    }

    private fun initCardHolderInput() {
        binding.textInputLayoutCardHolder.isVisible = delegate.componentParams.isHolderNameRequired
        binding.editTextCardHolder.setOnChangeListener {
            delegate.updateInputData { cardHolderName = binding.editTextCardHolder.rawValue }
            binding.textInputLayoutCardHolder.hideError()
        }

        binding.editTextCardHolder.onFocusChangeListener = OnFocusChangeListener { _, hasFocus ->
            val cardHolderValidation = delegate.outputData.cardHolderNameField.validation
            if (hasFocus) {
                binding.textInputLayoutCardHolder.hideError()
            } else if (cardHolderValidation is Validation.Invalid) {
                val errorReasonResId = cardHolderValidation.reason
                binding.textInputLayoutCardHolder.showError(localizedContext.getString(errorReasonResId))
            }
        }
    }

    private fun initStorePaymentMethodSwitch() {
        binding.switchStorePaymentMethod.isVisible = delegate.componentParams.isStorePaymentFieldVisible
        binding.switchStorePaymentMethod.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            delegate.updateInputData { isStorePaymentSelected = isChecked }
        }
    }

    override fun highlightValidationErrors() {
        val outputData = delegate.outputData

        var isErrorFocused = false
        val cardNumberValidation = outputData.cardNumberField.validation
        if (cardNumberValidation is Validation.Invalid) {
            isErrorFocused = true
            binding.editTextCardNumber.requestFocus()
            val errorReasonResId = cardNumberValidation.reason
            setCardNumberError(errorReasonResId)
        }

        val expiryFieldValidation = outputData.expiryDateField.validation
        if (expiryFieldValidation is Validation.Invalid) {
            if (!isErrorFocused) {
                binding.textInputLayoutExpiryDate.requestFocus()
            }
            val errorReasonResId = expiryFieldValidation.reason
            binding.textInputLayoutExpiryDate.showError(localizedContext.getString(errorReasonResId))
        }

        val cardHolderNameValidation = outputData.cardHolderNameField.validation
        if (cardHolderNameValidation is Validation.Invalid) {
            if (!isErrorFocused) {
                binding.textInputLayoutCardHolder.requestFocus()
            }
            val errorReasonResId = cardHolderNameValidation.reason
            binding.textInputLayoutCardHolder.showError(localizedContext.getString(errorReasonResId))
        }
    }

    private fun setCardNumberError(@StringRes stringResId: Int?) {
        if (stringResId == null) {
            binding.textInputLayoutCardNumber.hideError()
            binding.cardBrandLogoImageView.isVisible = true
        } else {
            binding.textInputLayoutCardNumber.showError(localizedContext.getString(stringResId))
            binding.cardBrandLogoImageView.isVisible = false
        }
    }

    override fun getView(): View = this
}
